Add EIP-5573: SIWE ReCap Extension (#5573)
ethereum/EIPs: EIPS/eip-4362.md

---
eip: 4362

> title: EIP-4361 Capabilities Extension (CapGrok)

Reviewer: ```suggestion
title: Sign-In with Ethereum Capabilities
```
Don't use `:` in titles. It causes inconsistency with jekyll (since it tries to interpret it as a key named `title: Sign-In with Ethereum` and a value of `Capabilities`.

CC @SamWilsn so you know for future reference.

## Motivation

SIWE CapGroks unlock integration of protocols and/or APIs for developers by reducing user friction, onchain state and increasing security by introducing informed consent and deterministic capability objects on top of Sign-In With Ethereum (EIP-4361).

## Specification

This specification has three different audiences:
- Web3 application developers that want to integrate CapGroks to authenticate with any protocols and APIs that support object capabilities.
- Protocol or API developers that want to learn how to define their own CapGroks.
- Wallet implementers that want to improve the UI for CapGroks.

### Terms and Definitions

> - CapGrok - A SIWE Message complying with this specification, i.e., containing at least one CapGrok URI in the `Resources` section and the corresponding human-readable CapGrok Statement appended to the SIWE `statement`.

Reviewer 2: I would propose the following breakdown for clarity:
* A CapGrok (or CapGrok Tuple if you want to be cheeky) expresses a capability consisting of a CapGrok URI + a CapGrok Statement (1:1 correspondence); optionally, it can be encapsulated in a CapGrok Details Object.
* A CapGrok Message is a SIWE message (string) that contains at least one CapGrok

I'm a little confused

- CapGrok URI - A type of URI under a certain namespace that resolves to a CapGrok Details Object.
- CapGrok Details Object - A JSON object describing the actions and optionally the resources associated with a CapGrok Capability under a certain namespace.
- Resource Service (RS) - The entity that is providing third-party services for the Ethereum account.
- SIWE Client (SC) - The entity initiating the SIWE authentication and CapGrok flow.
- Relying Party (RP) - same as SC in the context of authentication.

### Overview

This specification defines the following:
- CapGrok SIWE Extension

> - CapGrok Capability

Reviewer 2: I'd maybe rename that section "CapGrok Syntax" because you've already defined that a CapGrok is a Capability expression syntax?

   - CapGrok URI Scheme
   - CapGrok Details Object Schema
- CapGrok Translation Algorithm
- CapGrok Verification

### CapGrok SIWE Extension

> A CapGrok is a EIP-4361 message following a specific format that allows an Ethereum account to delegate a set of CapGrok Capabilities to a delegee through informed consent. Each CapGrok Capability MUST be represented by an entry in the `Resources` array of the SIWE message that MUST deterministically translate the CapGrok Capability in human-readable form to the `statement` field in the SIWE message using the CapGrok Translation Algorithm.

Reviewer 2: ```suggestion
A CapGrok Message is a EIP-4361 message following a specific format that allows an Ethereum account to delegate a set of CapGrok Capabilities to a delegee through informed consent. Each CapGrok Capability MUST be represented by an entry in the `Resources` array of the SIWE message that MUST deterministically translate the CapGrok Capability in human-readable form to the `statement` field in the SIWE message using the CapGrok Translation Algorithm.
```

The following SIWE message fields are used to further define (or limit) the scope of all CapGrok Capabilities:

> - The `URI` field MUST specify the delegee as a URI, e.g., `https://example.com`, `did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK`. It is expected that the RS authenticates the delegee before invoking an action for the CapGrok Capability.

Reviewer 3: EIP-4361 defines this field as a valid URI. Do we need to restate that requirement here?

Author: @SamWilsn I agree. I updated the PR accordingly.

Author: @bumblefudge added "intended delegee". Authentication is out-of-scope and is up to the RS.

- The `Issued At` field MUST be used to specify the issuance date of the CapGrok Capabilities.

> - If present, the `Expiration Time` field MUST be used as the expiration time of the CapGrok Capabilities where the RS no longer accepts invoking the associated actions.

Reviewer 2: ```suggestion
- If present, the `Expiration Time` field MUST be used as the expiration time of the CapGrok Capabilities, i.e. the time at which the RS will no longer accept an invocation of the capabilities expressed.
```

> - If present, the `Not Before` field MUST be used as the time that has to expire before the RS starts accepting invoking of the associated actions of the CapGrok Capabilities.

Reviewer 2: ```suggestion
- If present, the `Not Before` field MUST be used as the time that has to expire before the RS starts accepting invocations of the capabilities expressed in the message.
```

The following is a non-normative example of a SIWE message with the SIWE CapGrok Extension:
```text
example.com wants you to sign in with your Ethereum account:
0x0000000000000000000000000000000000000000

I further authorize https://example.com to perform the following actions on my behalf: (1) example: read for any. (2) example: append, delete for my.resource.1. (3) example: append for my.resource.2, my.resource.3.

URI: https://example.com
Version: 1
Chain ID: 1
Nonce: n-0S6_WzA2Mj
Issued At: 2022-06-21T12:00:00.000Z
Resources:
- urn:capability:example:eyJkZWYiOlsicmVhZCJdLCJ0YXIiOnsibXkucmVzb3VyY2UuMSI6WyJhcHBlbmQiLCJkZWxldGUiXSwibXkucmVzb3VyY2UuMiI6WyJhcHBlbmQiXSwibXkucmVzb3VyY2UuMyI6WyJhcHBlbmQiXX19
```

#### CapGrok Capability

A CapGrok Capability is identified by their CapGrok URI that resolves to a CapGrok Details Object which defines the associated actions and optional target resources. The scope of each CapGrok Capability is attenuated by common fields in the SWIE message as described in the previous chapter, e.g., `URI`, `Issued At`, `Expiration Time`, `Not Before`.

##### CapGrok URI Scheme

> A CapGrok URI starts with `urn:capability:` followed by the namespace discriminator, followed by `:` and the base64url-encoded payload of the CapGrok Details Object.

Reviewer 3: What is `base64url`? You _cannot_ link to its definition directly, but it might be worthwhile either including a copy of the spec (if it has a compatible license) in the `/assets/eip-5573` folder.

Author: @SamWilsn It is defined in RFC4648 --> chapter 5 defines base64url. I updated the PR to mention where base64url is defined.

Author: @SamWilsn pls let me know if this works for you.

The following is a non-normative example of a CapGrok Capability that uses the `example` namespace:
```text
urn:capability:example:eyJkZWZhdWx0QWN0aW9ucyI6WyJyZWFkIl0sInRhcmdldGVkQWN0aW9ucyI6eyJteS5yZXNvdXJjZS4xIjpbImFwcGVuZCIsImRlbGV0ZSJdLCJteS5yZXNvdXJjZS4yIjpbImFwcGVuZCJdLCJteS5yZXNvdXJjZS4zIjpbImFwcGVuZCJdfX0
```

> It is expected that RS implementers define their own namespace, e.g., `urn:capability:service:`.

Reviewer 2: Any registry or collision-protection expected? If not, perhaps the expectation could be more explicit, i.e. "RS implementers define a unique namespace per entity or per service".  If human-readability is expected (or detrimental), or if there are space constraints/length recommendations, I'd mention them.  Technically URNs can go to 4096 characters but I think that in practice, there are shorter maximums if you expect them to fit in, e.g., HTTP headers (this has come up at CASA for CAIP-10s)

Author: We don't expect that there is a registry needed since the RP and the RS know what they are talking about.

##### CapGrok Details Object Schema

> The CapGrok Details Object denotes which actions on which resources the delegee is authorized to invoke on behalf of the delegee for the validity period defined in the SIWE message. It can also contain additional information that the RS may require to verify a capability invocation. A CapGrok Details Object MUST follow the following JSON Schema:

Reviewer 2: ```suggestion
The CapGrok Details Object denotes which actions the delegee is authorized to take on which resources (in 1 or more action-resource tuples); it also denotes a validity period for the authorization(s), expressed using the standard validity mechanisms of the SIWE message. It can also contain additional information that the RS may require to verify a capability invocation. A CapGrok Details Object MUST follow the following JSON Schema:
```

The "additional information" is a little vague and open-ended.  Does this have to do with how the RS authenticated the delegee, or is that completely out of scope?  If in scope, maybe through an "i.e...." in there?

Author: addressed this

- Let `capgrok-transformed-statement` be the result of performing the above `CapGrok Translation Algorithm` with `uri` and `capgrok-uris` as input.
- Assert that the statement field of `capgrok-siwe` ends with `capgrok-transformed-statement`.

### Implementer's Guide

TBD

#### Web3 Application Implementers

TBD

#### Wallet Implementers

TBD

#### Protocol or API Implementers

TBD

## Rationale

While SIWE focuses on authenticating the Ethereum account against the service (relying party or SIWE client) initiating the SIWE flow, there is no canonical way to interact with a third-party service (resource service) on behalf of the authenticated Ethereum account. For example, a relying party might want to interact with another service on behalf of the Ethereum account, for example a service that provides data storage for the Ethereum account. This specification introduces a mechanism, that allows the service (or more generally a delegee) to combine authentication and authorization of such while preserving security and optimizing UX.

> Note, this approach is a similar mechanism to combining OpenID Connect (SIWE auth) and OAuth2 (SIWE CapGrok) whereas SIWE CapGrok follows an Object Capability-based approach.

Reviewer 3: Your Rationale is describing why the EIP as a whole should exist. This type of content belongs in the Motivation section.

The Rationale is meant to describe individual technical choices in the EIP itself.

Author: I agree, I moved it and I will provide contents for the Rationale section later.

## Security Considerations

> Resource service implementer's should not consider CapGroks as bearer tokens but instead require to authenticate the delegee in addition. The process of authenticating the delegee against the resource service is out of scope of this specification and can be done in various different ways.

Reviewer 3: As someone not familiar with authentication systems, I find the first sentence unnerving. Would you mind expanding it a bit, and perhaps explain mitigations or provide examples of what "require to authenticate the delegee in addition" means?

---
eip: 4362

> title: SIWE CapGrok Extension

Reviewer: ```suggestion
title: EIP-4361 Capabilities Extension
```
Slightly more descriptive title.

Author: Thanks for the recommendation! I changed the title to EIP-4361 Capabilities Extension (CapGrok).

Author: Does this sound ok to you @Pandapip1 ?

description: A mechanism on top of Sign-In With Ethereum for informed consent to delegate capabilities with an extensible scope mechanism.
author: Oliver Terbu (@awoie), Jacob Ward (@cobward), Charles Lehner (@clehner), Sam Gbafa (@skgbafa), Wayne Chang (@wyc)
discussions-to: https://ethereum-magicians.org/t/eip-4362-siwe-capgrok/10627
status: Draft
type: Standards Track
category: ERC
created: 2021-07-20
requires: 4361
---

## Abstract
Sign-In with Ethereum (SIWE) describes how Ethereum accounts authenticate with off-chain services (relying parties) by signing a standard message format parameterized by scope, session details, and security mechanisms (e.g., a nonce). This specification describes a mechanism on top of SIWE to give informed consent to delegate capabilities with a certain extensible scope mechanism to an authorized delegee. The way how a delegee authenticates against the target resource is out of scope of this specification and depends on the implementation of the target resource. This specification is fully compatible with [EIP-4361](./eip-4361.md).

## Motivation

> (non-normative)
> 

Reviewer: ```suggestion
```

SIWE CapGroks unlock integration of protocols and/or APIs for developers by reducing user friction, onchain state and increasing security by introducing informed consent and deterministic capability objects on top of Sign-In With Ethereum (EIP-4361).

## Specification

This specification has three different audiences:
- Web3 application developers that want to integrate CapGroks to authenticate with any protocols and APIs that support object capabilities.
- Protocol or API developers that want to learn how to define their own CapGroks.
- Wallet implementers that want to improve the UI for CapGroks.

### Terms and Definitions

> (non-normative)
> 

Reviewer: ```suggestion
```
Also, this is part of the specification, so it absolutely is normative.

The following SIWE message fields are used to further define (or limit) the scope of all CapGrok Capabilities:
- The `URI` field MUST specify the delegee as a URI, e.g., `https://example.com`, `did:key:z6MkhaXgBZDvotDkL5257faiztiGiC2QtKLGpbnnEGta2doK`. It is expected that the RS authenticates the delegee before invoking an action for the CapGrok Capability.
- The `Issued At` field MUST be used to specify the issuance date of the CapGrok Capabilities.
- If present, the `Expiration Time` field MUST be used as the expiration time of the CapGrok Capabilities where the RS no longer accepts invoking the associated actions.
- If present, the `Not Before` field MUST be used as the time that has to expire before the RS starts accepting invoking of the associated actions of the CapGrok Capabilities.

The following is a non-normative example of a SIWE message with the SIWE CapGrok Extension:

> ```

Reviewer: ```suggestion
```text
```
Please specify languages

I further authorize https://example.com to perform the following actions on my behalf: (1) example: read for any. (2) example: append, delete for my.resource.1. (3) example: append for my.resource.2, my.resource.3.

URI: https://example.com
Version: 1
Chain ID: 1
Nonce: n-0S6_WzA2Mj
Issued At: 2022-06-21T12:00:00.000Z
Resources:
- urn:capability:example:eyJkZWYiOlsicmVhZCJdLCJ0YXIiOnsibXkucmVzb3VyY2UuMSI6WyJhcHBlbmQiLCJkZWxldGUiXSwibXkucmVzb3VyY2UuMiI6WyJhcHBlbmQiXSwibXkucmVzb3VyY2UuMyI6WyJhcHBlbmQiXX19
```

#### CapGrok Capability

A CapGrok Capability is identified by their CapGrok URI that resolves to a CapGrok Details Object which defines the associated actions and optional target resources. The scope of each CapGrok Capability is attenuated by common fields in the SWIE message as described in the previous chapter, e.g., `URI`, `Issued At`, `Expiration Time`, `Not Before`.

##### CapGrok URI Scheme

A CapGrok URI starts with `urn:capability:` followed by the namespace discriminator, followed by `:` and the base64url-encoded payload of the CapGrok Details Object.

The following is a non-normative example of a CapGrok Capability that uses the `example` namespace:

> ```

Reviewer: ```suggestion
```text
```

urn:capability:example:eyJkZWZhdWx0QWN0aW9ucyI6WyJyZWFkIl0sInRhcmdldGVkQWN0aW9ucyI6eyJteS5yZXNvdXJjZS4xIjpbImFwcGVuZCIsImRlbGV0ZSJdLCJteS5yZXNvdXJjZS4yIjpbImFwcGVuZCJdLCJteS5yZXNvdXJjZS4zIjpbImFwcGVuZCJdfX0
```

It is expected that RS implementers define their own namespace, e.g., `urn:capability:service:`.

##### CapGrok Details Object Schema

The CapGrok Details Object denotes which actions on which resources the delegee is authorized to invoke on behalf of the delegee for the validity period defined in the SIWE message. It can also contain additional information that the RS may require to verify a capability invocation. A CapGrok Details Object MUST follow the following JSON Schema:

> ```

Reviewer: ```suggestion
```jsonc
```

            "type": "string",
            "minLength": 1
          },  
          "minItems": 1
        }   
      },  
      "additionalProperties": false,
      "minProperties": 1
    },  
    "ext": {
      "type": "object",
      "minProperties": 1
    }   
  },  
  "minProperties": 1,
  "additionalProperties": false,
  "dependentSchemas": {
    "ext": {
      "minProperties": 2
    }   
  }
}
```

A CapGrok Details Object defines the following properties:
- `def`: (CONDITIONAL) If present, `def` MUST be a JSON array of string values with at least one entry where each value describes an action the delegee MAY invoke in the RS on behalf of the Ethereum account without tying the scope to a particular target.
- `tar`: (CONDITIONAL) If present, `tar` MUST be a JSON object with variable properties where each property is a JSON array of string values each describing an action the delegee MAY invoke in the RS on behalf of the Ethereum account on the target resource denoted by the property name.
- `ext`: (OPTIONAL) If present, `ext` MUST be a JSON object with variable properties.

The following is a non-normative example of a CapGrok Capability Object with `def`, `tar` and `ext`:

> ```

Reviewer: ```suggestion
```jsonc
```

      ],
      "my.resource.2":[
         "append"
      ],
      "my.resource.3":[
         "append"
      ]
   },
   "ext":{
       "parentCapability": "bafybeigk7ly3pog6uupxku3b6bubirr434ib6tfaymvox6gotaaaaaaaaa"
   }
}
```

In the example above, the delegee is authorized to perform the action `read` independent of any resource, `append`, `delete` on resource `my.resource.1`, `append` on resource `my.resource.2` and `append` on `my.resource.3`. Note, the delegee can invoke each action invididually and independent from each other in the RS. Additionally the CapGrok Capability Object contains some additional information that the RS will need during verification. The responsibility for defining the structure and semantics of this data lies with the RS.

It is expected that RS implementers define which resources they want to expose through CapGrok Details Objects and which actions they want to allow to invoke on them.

#### CapGrok Translation Algorithm

After applying the CapGrok Translation Algorithm on a given SIWE message that MAY include a pre-defined `statement`, the `capgrok-transformed-statement` in a CapGrok SIWE message MUST conform to the following ABNF:

> ```

Reviewer: ```suggestion
```text
```

      - Append `namespace` concatenated with `": "` to `capgrok-transformed-statement`.
      - Append `actions` to `capgrok-transformed-statement`.
      - Append the string `" for "` to `capgrok-transformed-statement`.
      - Append `targets` to `capgrok-transformed-statement`.
      - Append the string `" ."` to `capgrok-transformed-statement`.
      - Increase `numbering` by 1.
- Return `capgrok-transformed-statement`.

#### CapGrok Verification

The following algorithm or an algorithm that produces the same output MUST be performed to verify a SIWE CapGrok.

Inputs:
- Let `capgrok-siwe` be the input SIWE message conforming to EIP-4361 and this EIP.
Algorithm:
- Let `uri` be the uri field of `capgrok-siwe`.
- Let `capgrok-uris` be an array of CapGrok URIs taken in order from the resources field of `capgrok-siwe`, such that URIs which are not valid CapGrok URIs are ignored.
- Let `capgrok-transformed-statement` be the result of performing the above `CapGrok Translation Algorithm` with `uri` and `capgrok-uris` as input.
- Assert that the statement field of `capgrok-siwe` ends with `capgrok-transformed-statement`.

### Implementer's Guide

> 
> (non-normative)

Reviewer: ```suggestion

TBD
```
Required: Remove that non-normative thing. It's part of the spec, so it is normative..

#### Web3 Application Implementers

TBD

#### Wallet Implementers

TBD

#### Protocol or API Implementers

TBD

## Rationale

While SIWE focuses on authenticating the Ethereum account against the service (relying party or SIWE client) initiating the SIWE flow, there is no canonical way to interact with a third-party service (resource service) on behalf of the authenticated Ethereum account. For example, a relying party might want to interact with another service on behalf of the Ethereum account, for example a service that provides data storage for the Ethereum account. This specification introduces a mechanism, that allows the service (or more generally a delegee) to combine authentication and authorization of such while preserving security and optimizing UX.

Note, this approach is a similar mechanism to combining OpenID Connect (SIWE auth) and OAuth2 (SIWE CapGrok) whereas SIWE CapGrok follows an Object Capability-based approach.

> ## Reference Implementation
> 
> - https://github.com/spruceid/capgrok

Reviewer: ```suggestion
```
Required: Must be put in the assets folder or deleted.

## Security Considerations

Resource service implementer's should not consider CapGroks as bearer tokens but instead require to authenticate the delegee in addition. The process of authenticating the delegee against the resource service is out of scope of this specification and can be done in various different ways.

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```
